iroha_p2p/src/peer.rs

```diff
@@ -0,0 +1,327 @@
+                                ctx.notify_with(self.stream());
+                            },
+                            _ => {
+                                warn!("Error reading server hello!");
+                                self.state = State::Error;
+                            }
+                        }
+                    }
+                }
+                Err(e) => {
+                    warn!("Could not connect to peer on {}! {}", addr, e);
+                }
+            }
+        }
+    }
+}
+
+#[derive(Copy, Clone)]
+pub enum State {
+    Connecting,
+    Connected,
+    ClientHelloSent,
+    ServerHelloSent,
+    Ready,
+    Error
+}
+
+pub async fn readable<T: AsRef<TcpStream>>(stream: T) -> io::Result<()> {
+    stream.as_ref().readable().await
+}
+
+pub async fn read_client_hello(stream: &mut OwnedReadHalf) -> Result<Message, Error> {
+    let size = stream.read_u8().await? as usize;
+    if size < 256 {
+        // Reading garbage
+        let mut buf = vec![0u8; size];
+        stream.read_exact(&mut buf).await?;
+    } else {
+        return Err(io::Error::from(ErrorKind::InvalidInput));
+    }
+    // And then we have clients public key
+    let mut buf = [0u8; 32];
+    stream.read_exact(&mut buf).await?;
+    Ok(Message::client_hello(buf))
+}
+
+pub async fn send_client_hello(stream: &mut OwnedWriteHalf, key: &[u8]) -> io::Result<()> {
+    let garbage = generate_garbage();
+
+    stream.write_u8(garbage.len() as u8).await?;
+    stream.write_all(&garbage).await?;
+    stream.write_all(&key).await?;
+
+    Ok(())
+}
+
+pub async fn read_server_hello(stream: &mut OwnedReadHalf) -> Result<Message, Error> {
+    let size = stream.read_u8().await? as usize;
+    if size < 256 {
+        // Reading garbage
+        let mut buf = vec![0u8; size];
+        stream.read_exact(&mut buf).await?;
+    } else {
+        return Err(io::Error::from(ErrorKind::InvalidInput));
+    }
+    // Then we have clients public key
+    let mut key = [0u8; 32];
+    stream.read_exact(&mut key).await?;
+    // Then the nonce
+    let mut nonce = [0u8; 12];
+    stream.read_exact(&mut nonce).await?;
+    Ok(Message::server_hello(key, nonce))
+}
+
+pub async fn send_server_hello(stream: &mut OwnedWriteHalf, key: &[u8], nonce: &[u8]) -> io::Result<()> {
+    let garbage = generate_garbage();
+
+    stream.write_u8(garbage.len() as u8).await?;
+    stream.write_all(&garbage).await?;
+    stream.write_all(&key).await?;
+    stream.write_all(&nonce).await?;
+
+    Ok(())
+}
+
+pub async fn read_message(stream: &mut OwnedReadHalf) -> Result<Message, Error> {
+    let size = stream.read_u32().await? as usize;
+    if size > 0 && size < MAX_MESSAGE_LENGTH {
+        let mut buf = vec![0u8; size];
+        let mut read = 0;
+        while read < size {
+            read += stream.read_exact(&mut buf[read..]).await?;
+        }
+
+        return Ok(Message::data(&buf));
+    }
+    Err(io::Error::from(ErrorKind::InvalidInput))
```

Comment:
It is better not to have nestings like that. You could check some predicate over input and return error on fail. It will remove nesting level

Reply:
Fixed.